File: Project_Nr_1/Part_1/Receiver/PayloadHandler.cs
using Message_Agent.Common;
using Newtonsoft.Json;
using System.Text;
using System.Collections.Generic;

namespace Receiver
{
    class PayloadHandler
    {
        private static Dictionary<string, List<string>> _messages =
            new Dictionary<string, List<string>>();

        private static object _locker = new object();

        public static void Handle(byte[] payloadBytes)
        {
            var payloadString = Encoding.UTF8.GetString(payloadBytes);
            var payload = JsonConvert.DeserializeObject<PayLoad>(payloadString);

            if (payload == null)
            {
                return;
            }

            lock (_locker)
            {
                if (!_messages.ContainsKey(payload.Topic))
                {
                    _messages[payload.Topic] = new List<string>();
                }

                _messages[payload.Topic].Add(payload.Message);
            }
        }

        public static Dictionary<string, List<string>> GetMessages()
        {
            lock (_locker)
            {
                return new Dictionary<string, List<string>>(_messages);
            }
        }
    }
}

File: Project_Nr_1/Part_1/Python_Receiver/payload_handler.py
import json
import threading

class PayloadHandler:
    _messages = {}
    _lock = threading.Lock()

    @staticmethod
    def handle(payload_bytes):
        payload_string = payload_bytes.decode("utf-8")

        try:
            payload = json.loads(payload_string)
        except json.JSONDecodeError:
            return

        if payload is None:
            return

        topic = payload.get("Topic")
        message = payload.get("Message")

        if topic is None or message is None:
            return

        with PayloadHandler._lock:
            if topic not in PayloadHandler._messages:
                PayloadHandler._messages[topic] = []

            PayloadHandler._messages[topic].append(message)

    @staticmethod
    def get_messages():
        with PayloadHandler._lock:
            return {
                topic: messages.copy()
                for topic, messages in PayloadHandler._messages.items() 
            }